workflow/controller/workflowpod.go

```diff
@@ -695,83 +704,98 @@ func addSchedulingConstraints(pod *apiv1.Pod, wfSpec *wfv1.WorkflowSpec, tmpl *w
 	}
 }
 
-// addVolumeReferences adds any volumeMounts that a container/sidecar is referencing, to the pod.spec.volumes
-// These are either specified in the workflow.spec.volumes or the workflow.spec.volumeClaimTemplate section
-func addVolumeReferences(pod *apiv1.Pod, vols []apiv1.Volume, tmpl *wfv1.Template, pvcs []apiv1.Volume) error {
-	switch tmpl.GetType() {
-	case wfv1.TemplateTypeContainer, wfv1.TemplateTypeScript:
-	default:
-		return nil
-	}
-
+func (woc *wfOperationCtx) addVolumeRef(volMounts []apiv1.VolumeMount, pod *apiv1.Pod, tmpl *wfv1.Template) error {
 	// getVolByName is a helper to retrieve a volume by its name, either from the volumes or claims section
-	getVolByName := func(name string) *apiv1.Volume {
+	getVolByName := func(name string, tmpl *wfv1.Template) *apiv1.Volume {
 		// Find a volume from template-local volumes.
 		for _, vol := range tmpl.Volumes {
 			if vol.Name == name {
 				return &vol
 			}
 		}
 		// Find a volume from global volumes.
-		for _, vol := range vols {
+		for _, vol := range woc.volumes {
 			if vol.Name == name {
 				return &vol
 			}
 		}
 		// Find a volume from pvcs.
-		for _, pvc := range pvcs {
+		for _, pvc := range woc.wf.Status.PersistentVolumeClaims {
 			if pvc.Name == name {
 				return &pvc
 			}
 		}
 		return nil
 	}
 
-	addVolumeRef := func(volMounts []apiv1.VolumeMount) error {
-		for _, volMnt := range volMounts {
-			vol := getVolByName(volMnt.Name)
-			if vol == nil {
-				return errors.Errorf(errors.CodeBadRequest, "volume '%s' not found in workflow spec", volMnt.Name)
-			}
-			found := false
-			for _, v := range pod.Spec.Volumes {
-				if v.Name == vol.Name {
-					found = true
-					break
-				}
+	for _, volMnt := range volMounts {
+		vol := getVolByName(volMnt.Name, tmpl)
+		if vol == nil {
+			return errors.Errorf(errors.CodeBadRequest, "volume '%s' not found in workflow spec", volMnt.Name)
```

Comment:
is this a bad request?

Reply:
Yes, if the volume should be defined in workflow otherwise it is an invalid workflow spec.